### PR TITLE
Fixed ClusterShardingMessageSerializer

### PR DESCRIPTION
**Pull request by alexvaluyskiy**
Fixed ClusterShardingMessageSerializer and added ClusterShardingMessageSerializerSpec

### CODE REVIEW

File: src/contrib/cluster/Akka.Cluster.Sharding/PersistentShardCoordinator.cs [hunk 1]
```diff
@@ -133,6 +134,53 @@ public State Updated(IDomainEvent e)
 
                 return new State(shards ?? Shards, regions ?? Regions, regionProxies ?? RegionProxies, unallocatedShards ?? UnallocatedShards);
             }
+
+            #region Equals
+
+            public override bool Equals(object obj)
```
Comment (alexvaluyskiy) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShardCoordinator.cs]: Not the best my equality comparer. We should compare objects in a different way

File: src/contrib/cluster/Akka.Cluster.Sharding/PersistentShardCoordinator.cs [hunk 2]
```diff
@@ -133,6 +134,53 @@ public State Updated(IDomainEvent e)
...
+                if (ReferenceEquals(other, null)) return false;
+                if (ReferenceEquals(other, this)) return true;
+
+                return Shards.SequenceEqual(other.Shards)
+                    && Regions.Keys.SequenceEqual(other.Regions.Keys)
+                    && RegionProxies.SequenceEqual(other.RegionProxies)
+                    && UnallocatedShards.SequenceEqual(other.UnallocatedShards);
+            }
+
+            public override int GetHashCode()
```
Comment (alexvaluyskiy) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShardCoordinator.cs]: I don't like this GetHashCode

Comment (Danthar) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShardCoordinator.cs]: The fixed integer values are a smell, and a problem if the same gethashcode is copy pasted everywhere. But other then that its not like we need encryptionlevel hashcodes here.

**Comment by Danthar** [comment]
needs rebase.

**Pull request merged**